jax/experimental/jax2tf/tests/primitive_harness.py

```diff
@@ -179,6 +180,212 @@ def parameterized(harness_group: Iterable[Harness],
+  ]
+)
+
+def _make_sub_harness(name, *, lhs_shape=(2, 3), rhs_shape=(2, 3),
+                      dtype=np.float32):
+  return Harness(f"{name}_lhs={jtu.format_shape_dtype_string(lhs_shape, dtype)}_rhs={jtu.format_shape_dtype_string(rhs_shape, dtype)}",
+                 lax.sub,
+                 [RandArg(lhs_shape, dtype), RandArg(rhs_shape, dtype)],
+                 lhs_shape=lhs_shape,
+                 rhs_shape=rhs_shape,
+                 dtype=dtype)
+
+lax_sub = tuple( # Validate dtypes
+  _make_sub_harness("dtypes", dtype=dtype)
+  for dtype in set(jtu.dtypes.all) - set(jtu.dtypes.boolean)
+) + tuple( # Validate broadcasting behaviour
+  _make_sub_harness("broadcasting", lhs_shape=lhs_shape, rhs_shape=rhs_shape)
+  for lhs_shape, rhs_shape in [
+    ((), (2, 3)),     # broadcast scalar
+    ((1, 2), (3, 2)), # broadcast along specific axis
+  ]
+)
+
+def _make_bitcast_convert_type_harness(name, *, shape=(2, 3), dtype=np.float32,
+                                       new_dtype=np.float32):
+  return Harness(f"{name}_shape={jtu.format_shape_dtype_string(shape, dtype)}_newdtype={np.dtype(new_dtype).name}",
+                 lambda x: (
+                     lax.bitcast_convert_type_p.bind(x, new_dtype=new_dtype)),
+                 [RandArg(shape, dtype)],
+                 shape=shape,
+                 dtype=dtype,
+                 new_dtype=new_dtype)
+
+def _can_bitcast(dtype, target_dtype):
+  def _to_equivalence_class(dtype):
+    if dtypes.issubdtype(dtype, np.integer): return dtypes.iinfo(dtype).bits
+    elif dtypes.issubdtype(dtype, np.floating): return dtypes.finfo(dtype).bits
+    else:
+      assert dtype == np.bool_ or dtypes.issubdtype(dtype, np.complexfloating)
+      # Complex and boolean types can only be cast to themselves
+      return np.dtype(dtype).name
+  return _to_equivalence_class(dtype) == _to_equivalence_class(target_dtype)
+
+lax_bitcast_convert_type = tuple( # Validate dtypes combinations
+  _make_bitcast_convert_type_harness("dtypes_to_new_dtypes", dtype=dtype,
+                                     new_dtype=new_dtype)
+  for dtype in jtu.dtypes.all
+  for new_dtype in filter(partial(_can_bitcast, dtype), jtu.dtypes.all)
+)
+
+def _make_add_jaxvals_harness(name, *, shapes=((2,), (2,)),
+                              dtype=np.float32):
+  return Harness(f"{name}_lhs={jtu.format_shape_dtype_string(shapes[0], dtype)}_rhs={jtu.format_shape_dtype_string(shapes[1], dtype)}",
+                 ad_util.add_jaxvals_p.bind,
+                 list(map(lambda s: RandArg(s, dtype), shapes)),
+                 shapes=shapes,
+                 dtypes=dtypes)
+
+ad_util_add_jaxvals = tuple( # Validate dtypes
+  _make_add_jaxvals_harness("dtypes", dtype=dtype)
+  for dtype in set(jtu.dtypes.all) - set(jtu.dtypes.boolean)
+)
+
+lax_tie_in = tuple(
+  Harness(f"lhs={jtu.format_shape_dtype_string(lhs_shape, lhs_dtype)}_rhs={jtu.format_shape_dtype_string(rhs_shape, rhs_dtype)}",
+          lax.tie_in_p.bind,
```

Comment:
This primitive is deprecated and not used if omnistaging is on. It does not hurt to test it though.

Reply:
The test is actually skipped if omnistaging is on. Since it is still possible to end up with a call to `tie_in_p` somewhere by disabling omnistaging, I figured it is best to ensure it works properly.